### PR TITLE
Submit transactions in bulk

### PR DESCRIPTION
**Pull request by ThibautBremand**
- Allow to submit a bulk of raw transaction payloads.
- Add a Permissions view to enable or disable the permission to submit transactions in bulk
- Use `noWait (boolean)` and `onError (string)` in the submitTransactionsBulk route to control if we want to wait for the tx hashes or not, and if we want to continue or abort when one tx fails
- Resolve NFTs in the submitTransactionBulk view for NFT related transactions (mint, burn, NFT offers)
- Docs: https://github.com/GemWallet/gemwallet-website/pull/40

### CODE REVIEW

File: packages/extension/src/utils/storageInMemory.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/*
+ * Manages the storage in memory
+ */
+
+export const saveInMemory = (key: string, value: string) => {
+  try {
```
Comment (ThibautBremand) [inline packages/extension/src/utils/storageInMemory.ts:6]: To be replaced by the lib introduced recently in master

File: packages/constants/src/payload/payload.types.ts [hunk 2]
```diff
@@ -240,6 +240,32 @@ export interface SubmitTransactionRequest {
...
+  onError?: 'abort' | 'continue'; // default: abort
+}
+
+// Type that goes through the messaging system. It embeds the transactions in a Record with a numbered key, in order
+// to guarantee that the order is preserved.
+export interface SubmitTransactionsBulkModifiedRequest {
+  // The key is used to guarantee that the transactions are submitted in the same order as they are in the request.
+  transactions: Record<number, TransactionWithID>;
+  onError?: 'abort' | 'continue'; // default: abort
+}
```
Comment (ThibautBremand) [inline packages/constants/src/payload/payload.types.ts:263]: noWait missing
Need to do some inheritance here

File: packages/extension/src/utils/NFTImageResolver.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import { convertHexToString } from 'xrpl';
+
+import { parseJSON } from './NFTViewer';
+
+export interface NFTData {
+  NFTokenID: string;
+  NFType?: string;
```
Comment (ThibautBremand) [inline packages/extension/src/utils/NFTImageResolver.ts:7]: nftType (check out the xls standard)

File: packages/extension/src/utils/parseFromString.ts [hunk 4]
```diff
@@ -379,3 +380,18 @@ export const parseTransactionParam = (str: string | null): Transaction | null =>
 
   return null;
 };
+
+export const parseTransactionsBulkMap = (
```
Comment (ThibautBremand) [inline packages/extension/src/utils/parseFromString.ts:384]: Missing unit tests

File: packages/extension/src/contexts/LedgerContext/LedgerContext.tsx [hunk 5]
```diff
@@ -751,6 +776,88 @@ const LedgerProvider: FC = ({ children }) => {
     [client, getCurrentWallet]
   );
 
+  const submitTransactionsBulk = useCallback(
+    async (payload: SubmitTransactionsBulkRequest) => {
+      const onError = payload.onError;
```
Comment (florent-uzio) [inline packages/extension/src/contexts/LedgerContext/LedgerContext.tsx:781]: You could destructure here

```jsx
const { onError, transactions } = payload;
```

File: packages/extension/src/contexts/LedgerContext/LedgerContext.tsx [hunk 6]
```diff
@@ -751,6 +776,88 @@ const LedgerProvider: FC = ({ children }) => {
...
+          Sentry.captureException(e);
+          return {
+            ID,
+            error: `Error while processing the transaction: ${JSON.stringify(e)}`
+          };
+        }
+      };
+
+      const results: TransactionBulkResponse[] = [];
+      for (const tx of payload.transactions) {
```
Comment (florent-uzio) [inline packages/extension/src/contexts/LedgerContext/LedgerContext.tsx:845]: This loop will send the requests one by one which is slow.
Have you thought about using `Promise.all` or `Promise.allSettled()` to send them in parallel and make the whole process faster?

File: packages/extension/src/contexts/LedgerContext/LedgerContext.tsx [hunk 7]
```diff
@@ -768,6 +875,62 @@ const LedgerProvider: FC = ({ children }) => {
     }
   }, [client, getCurrentWallet]);
 
+  const getNFTInfo = useCallback(
+    async (NFTokenID: string, network?: string): Promise<NFTInfoResponse> => {
+      const connectToLedger = async (server: string): Promise<Client> => {
```
Comment (florent-uzio) [inline packages/extension/src/contexts/LedgerContext/LedgerContext.tsx:880]: I think this function to connect to the ledger could be extracted to its own file. Isn't there already a way to connect to the ledger (in a different file maybe?) ?

File: packages/extension/src/utils/storageChromeLocal.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Manages the storage in Chrome local storage
+ */
+
+export const saveInChromeLocalStorage = (key: string, value: any): Promise<void> => {
```
Comment (florent-uzio) [inline packages/extension/src/utils/storageChromeLocal.ts:5]: Why do you need it to return a promise?

What about something like this? You can use the helpers `isString`, `isNumber` etc. that I wrote in my previous PR.

```jsx
/**
 * Set an item (key/value) in local storage.
 *
 * @param {string} key The key
 * @param {unknown} value The value
 */
export const setLocalStorageItem = (
  key: string,
  value: Record<string, any> | string | number | boolean,
) => {
  try {
    if (isString(value)) {
      localStorage.setItem(key, value)
    } else if (isNumber(value) || isBoolean(value)) {
      localStorage.setItem(key, value.toString())
    } else if (isObject(value)) {
      localStorage.setItem(key, JSON.stringify(value))
    } else {
      console.error(`Unrecognized local storage value ${value} for key: ${key}`)
    }
  } catch (err) {
    console.error(err)
  }
}
```

File: packages/extension/src/utils/storageChromeLocal.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+          resolve();
+        }
+      });
+    } else {
+      resolve();
+    }
+  });
+};
+
+export const loadFromChromeLocalStorage = (
```
Comment (florent-uzio) [inline packages/extension/src/utils/storageChromeLocal.ts:21]: What about sth like this? You can also implement your delete logic

```jsx
/**
 * Retrieve an item from the local storage.
 *
 * @param {string} key
 * @returns Any value.
 */
export const getLocalStorageItem = <T = unknown>(
  key: string,
): T | undefined => {
  const value = localStorage.getItem(key)

  if (value === null || value === 'undefined') {
    return undefined
  }

  try {
    return JSON.parse(value) as T
  } catch (_err) {
    return value as unknown as T
  }
}
```

**Review by florent-uzio — COMMENTED** [review]
Nice job I left a few comments but I can have another look later. Let me know.

File: packages/extension/src/contexts/LedgerContext/LedgerContext.tsx (hunk 6 above)
Comment (ThibautBremand) [reply packages/extension/src/contexts/LedgerContext/LedgerContext.tsx:845]: I tried with `Promise.all` but it was only really submitting the first transaction of each batch. My assumption was that I needed to throttle

File: packages/constants/src/network/network.constant.ts [hunk 10]
```diff
@@ -13,6 +13,11 @@ export enum NetworkServer {
   AMM_DEVNET = 'wss://amm.devnet.rippletest.net:51233'
 }
 
+export enum MainnetClioServer {
```
[the remark this answers is not in the document]
Comment (ThibautBremand) [reply packages/constants/src/network/network.constant.ts]: It was to isolate Clio servers but yes we can integrate them into the original enum 👍

File: packages/extension/src/utils/storageChromeLocal.ts (hunk 8 above)
Comment (ThibautBremand) [reply packages/extension/src/utils/storageChromeLocal.ts:5]: Your code sample is for localStorage whici is synchronous but here we use Chrome.storage.local which is async by nature. I can see if the code can be changed, but keep in mind it is not the same storage

File: packages/constants/src/payload/payload.types.ts [hunk 11]
```diff
@@ -240,6 +240,39 @@ export interface SubmitTransactionRequest {
...
+
+export const DEFAULT_SUBMIT_TX_BULK_ON_ERROR = 'abort';
+
+export type TransactionErrorHandling = 'abort' | 'continue';
+
+export interface BaseTransactionBulkRequest {
+  // If set to true, the function will not wait for the transaction hashes to be returned from the XRPL network, and
+  // 'accepted' values will be returned for each transaction, instead of 'hash'.
+  // Default: false
+  noWait?: boolean;
```
[the remark this answers is not in the document]
Comment (ThibautBremand) [reply packages/constants/src/payload/payload.types.ts:257]: Or maybe `waitForHashes` (default: true)

File: packages/constants/src/payload/payload.types.ts [hunk 12]
```diff
@@ -244,6 +244,39 @@ export interface SignTransactionRequest {
...
+  // This id is only used as an indicator in the response, and is not used to order transactions.
+  ID?: string;
+};
+
+export const DEFAULT_SUBMIT_TX_BULK_ON_ERROR = 'abort';
+
+export type TransactionErrorHandling = 'abort' | 'continue';
+
+export interface BaseTransactionBulkRequest {
+  // If set to true, the function will not wait for the transaction hashes to be returned from the XRPL network, and
```
Comment (FlorianBouron) [inline packages/constants/src/payload/payload.types.ts:258]:
```suggestion
  // If set to true, the function will not wait for the transaction to be executed on XRPL
```

File: packages/api/templates/index.html [hunk 13]
```diff
@@ -554,704 +554,20 @@
...
-              TransactionType: 'Payment',
-              Destination: 'rhikRdkFw28csKw9z7fVoBjWncz1HSoQij',
-              Amount: '100000',
-              Memos: [
-                {
-                  Memo: {
-                    MemoData: '54657374206D656D6F',
-                    MemoType: '4465736372697074696F6E'
-                  }
+          const transactions = Array.from({ length: 50 }, (_, index) => ({
```
Comment (ThibautBremand) [inline packages/api/templates/index.html:557]: You removed the bored apes minting :( (and a setTrustline)

Comment (FlorianBouron) [reply packages/api/templates/index.html:557]: Oh I see, we can add this back tomorrow :)

**Review by FlorianBouron — APPROVED** [review]

**Pull request merged**